### ISSUE TITLE
Rebuilding fails to find include directory

### ISSUE THREAD

**Issue by yt87** [mesonbuild/meson-python#393]
I have trouble building my project with mesonpy 0.13 dev version. It worked with 0.12. The issue can be illustrated with the example `our_first_module` .
I added a single line to the .c file:
```
#include <numpy/ndarraytypes.h> 
```
I modified `pyproject.toml` to include dependency on numpy (which is already installed in my conda test environment and added include entries to `meson.build`:
```
project('purelib-and-platlib', 'c')

py_mod = import('python')
py = py_mod.find_installation(pure: false)

incdir_numpy = run_command(py,
  [
    '-c',
    'import numpy; print(numpy.get_include())'
  ],
  check: true
).stdout().strip()

py.extension_module(
    'our_first_module',
    'our_first_module.c',
    include_directories: [incdir_numpy],
    install: true,
)
```
Building in editable mode succeeded, file `.mesonpy/editable/build/our_first_module.cpython-311-x86_64-linux-gnu.so` is created
```
pip install -e . 
Obtaining file:///mnt/sdb1/projects/test
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Building wheels for collected packages: our-first-project
  Building editable for our-first-project (pyproject.toml) ... done
  Created wheel for our-first-project: filename=our_first_project-0.0.1-cp311-cp311-linux_x86_64.whl size=10492 sha256=081d3a2bd64bd078be6572836c753f4dcd6f8ecd00fd4433258040dfe5a442c6
  Stored in directory: /tmp/pip-ephem-wheel-cache-pwurv2t6/wheels/de/13/e3/83ca37426d2c8670b49e91bb81a7fe9434a99eeff17a1ca824
Successfully built our-first-project
Installing collected packages: our-first-project
Successfully installed our-first-project-0.0.1
```
However import fails:
```
>>> import our_first_module
+ rebuilding /mnt/sdb1/projects/test
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /mnt/sdb1/home/george/miniforge3/envs/apps/bin/ninja
[1/2] Compiling C object our_first_module.cpython-311-x86_64-linux-gnu.so.p/our_first_module.c.o
FAILED: our_first_module.cpython-311-x86_64-linux-gnu.so.p/our_first_module.c.o 
cc -Iour_first_module.cpython-311-x86_64-linux-gnu.so.p -I. -I../../.. -I/tmp/pip-build-env-8kytbnff/overlay/lib/python3.11/site-packages/numpy/core/include -I/mnt/sdb1/home/george/miniforge3/envs/apps/include/python3.11 -fvisibility=hidden -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O2 -fPIC -MD -MQ our_first_module.cpython-311-x86_64-linux-gnu.so.p/our_first_module.c.o -MF our_first_module.cpython-311-x86_64-linux-gnu.so.p/our_first_module.c.o.d -o our_first_module.cpython-311-x86_64-linux-gnu.so.p/our_first_module.c.o -c ../../../our_first_module.c
../../../our_first_module.c:2:10: fatal error: numpy/ndarraytypes.h: No such file or directory
    2 | #include <numpy/ndarraytypes.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1140, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1080, in _find_spec
  File "/mnt/sdb1/home/george/miniforge3/envs/apps/lib/python3.11/site-packages/_mesonpy_hook_our_first_project.py", line 130, in find_spec
    self.rebuild()
  File "/mnt/sdb1/home/george/miniforge3/envs/apps/lib/python3.11/site-packages/_mesonpy_hook_our_first_project.py", line 119, in rebuild
    self._proc(command)
  File "/mnt/sdb1/home/george/miniforge3/envs/apps/lib/python3.11/site-packages/_mesonpy_hook_our_first_project.py", line 111, in _proc
    subprocess.check_call(command, cwd=self._build_path, env=env)
  File "/mnt/sdb1/home/george/miniforge3/envs/apps/lib/python3.11/subprocess.py", line 413, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['meson', 'compile']' returned non-zero exit status 1.
```
So, I have 2 questions:
1. Why the module is rebuilt? 
2. The include path contains value of `incdir_numpy`, yet the include numpy header file is not found.

**Comment by yt87** [comment]
It appears the command to get numpy include returns the temporary meson build directory in `/tmp`. When I added `--no-clean` to the pip command line import worked fine:
```
$ pip install --no-clean --editable .  
Obtaining file:///mnt/sdb1/projects/test
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Building wheels for collected packages: our-first-project
  Building editable for our-first-project (pyproject.toml) ... done
  Created wheel for our-first-project: filename=our_first_project-0.0.1-cp311-cp311-linux_x86_64.whl size=10492 sha256=90fcbd2ae4aabe719eee51b3813ff2bedeb03191e723995312a752837b3d0f9e
  Stored in directory: /tmp/pip-ephem-wheel-cache-6k21y_ch/wheels/de/13/e3/83ca37426d2c8670b49e91bb81a7fe9434a99eeff17a1ca824
Successfully built our-first-project
Installing collected packages: our-first-project
Successfully installed our-first-project-0.0.1
$ python
Python 3.11.0 | packaged by conda-forge | (main, Jan 14 2023, 12:27:40) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import our_first_module
+ rebuilding /mnt/sdb1/projects/test
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /mnt/sdb1/home/george/miniforge3/envs/apps/bin/ninja
ninja: no work to do.
ninja: Entering directory `/mnt/sdb1/projects/test/.mesonpy/editable/build'
ninja: no work to do.
Installing our_first_module.cpython-311-x86_64-linux-gnu.so to /mnt/sdb1/projects/test/.mesonpy/editable/install/mnt/sdb1/home/george/miniforge3/envs/apps/lib/python3.11/site-packages
Skipping RPATH fixing
+ successfully rebuilt
>>> 
```
This does not appear to be a proper solution, I think.

**Comment by dnicolodi** [comment]
Things are working as expected. Unfortunately the interaction of different component in this case is less than ideal and not very intuitive. I've prepared some documentation (that targets the redesigned implementation of editable installs merged after meson-python 0.13.0dev0 was release, but it applied to the old implementation with only tiny caveats) that explains what is going on but it hasn't been merged yet. You can find it here #389.

pip uses an isolated build environment to build wheels, also when building editable installs. This means that it creates a throw away Python environment and installs in it the build dependencies for the package being built. In your case, the build dependencies include numpy. When `meson setup` is run by meson-python, it is run in this build environment, and it finds the numpy installation in it, and it records the location of the numpy headers as being in the throw away build environment. When the editable wheel is built and installed, the throw away build environment is, indeed, thrown away.

meson-python editable installations are designed to rebuild also extension modules (and in general anything that requires a compilation step) when the package installed in this mode is imported. Therefore, when you import your package, ninja tries to rebuild the extension module, but the include path used points to a directory that has been deleted in the meantime, and you get the error you observe.

The solution is to disable build isolation with the `--no-build-isolation` command line argument when doing editable installs (and to make sure that the build dependencies for your package are installed in the development environment):
```
python -m pip install --no-build-isolation --editable .
```

The `--no-clean` pip command line option works around the issue by not cleaning up the temporary build environment. However, as you observe, the temporary build environment is placed in `/tmp` and will be eventually be deleted. `--no-build-isolation` is the right solution.
